### ISSUE TITLE
How to know if rmarkdown runtime:shiny is being used or if inside a real Shiny app?

### ISSUE THREAD

**Issue by daattali** [rstudio/rmarkdown#2004]
If I want to check whether the current shiny server code is being run inside an rmarkdown or inside a shiny application, what's the best way to do so?

This old issue has been closed https://github.com/rstudio/rmarkdown/issues/933 and the solution there does not seem to work anymore. Solutions from [this SO thread](https://stackoverflow.com/questions/32806974/detecting-whether-shiny-runs-the-r-code) are not helpful either.

**Comment by daattali** [comment]
To clarify: querying for `knitr::opts_knit$get("rmarkdown.runtime")` or `rmarkdown::metadata$runtime` does return "shiny" when in Rmd and return "" when in shiny.

The problem is that this only works if the check is done directly in the authored file. If I develop a package that wants to check for this, then calling a function in a package from inside an rmarkdown no longer returns "shiny" for those checks.

**Comment by atusy** [comment]
You can use `rmarkdown::metadata`

````
---
title: "Untitled"
output: html_document
runtime: shiny
---

```{r}
rmarkdown::metadata
```

```{r, eval=FALSE}
library(shiny)
runApp(list(
  ui = bootstrapPage(
    shiny::verbatimTextOutput('metadata')
  ),
  server = function(input, output) {
    output$metadata <- shiny::renderText(capture.output(str(rmarkdown::metadata)))
  }
))
```
````

**Comment by daattali** [comment]
As mentioned in my comment above, I know this solution works for code inside the rmarkdown document itself.

But it doesn't work if placed inside other functions that are called from the rmarkdown doc. I'm developing a package that has funtions that can be called from a shiny app or an rmd, and the function in the package doesn't return anything for `rmarkdown::metadata`

**Comment by atusy** [comment]
Oops, sorry I missed it.
What about adding the argument to your function, and ask users to pass the `rmarkdown::metadata` to it?
Or possibly it works by specifying it as a default value owing to lazy evaluation (unsure).

```r
function(..., metadata = rmarkdown::metadata) {
}
```

**Comment by daattali** [comment]
Yes, adding an argument so that the user must explicitly say if it's rmd or not is possible. For best user experience, I'd like to do that automatically without requiring user input.

**Comment by cderv** [comment]
@daattali can you share an example where `knitr::opts_knit$get("rmarkdown.runtime")`does not work ? 

From what I know, this is the way to check it and this value of knitr's options should be available in the context of a R Markdown rendering even for external functions called. 

Here is a small example using an external function define in an R file that is sourced into the document.

````markdown
---
title: "Test"
output:
  html_document: default
runtime: shiny
---

```{cat, engine.opts = list(file = "test.R")}
is_shiny_in_rmd <- function() {
  knitr::opts_knit$get("rmarkdown.runtime")
}
```

```{r}
source("test.R")
```

```{r}
is_shiny_in_rmd()
```
````

resulting in 
![image](https://user-images.githubusercontent.com/6791940/104188654-d9cdf080-5419-11eb-9fae-a89d39536b04.png)


This is also how other package like **shiny** itself does it 
https://github.com/rstudio/shiny/blob/5b27d9258e8b18de15eda238a863bd239719870f/R/knitr.R#L15

See also a recherche for such code: 
https://github.com/search?q=knitr%3A%3Aopts_knit%24get%28%22rmarkdown.runtime%22%29&type=code

If that does not work for you, can you share more about it ? 

Thank you.

**Comment by daattali** [comment]
It took a few hours but I was finally able to track down a small isolated example of where this breaks.

The following code prints the runtime to a file twice: once when the document initializes and once when pressing a button from inside observeEvent. The first time it thinks it's inside an rmd, the second time it thinks it's in a shiny app.

````
---
output: html_document
runtime: shiny
---

```{r}
get_runtime <- function() {
  runtime <- knitr::opts_knit$get("rmarkdown.runtime")
  if (!is.null(runtime) && runtime == "shiny") {
    write("rmd", file = "test.txt", append = TRUE)
  } else {
    write("shiny", file = "test.txt", append = TRUE)
  }
}

get_runtime()
actionButton('click', 'click')

observeEvent(input$click, {
  get_runtime()
})
```
````

**Comment by cderv** [comment]
This is a specific example that reproduce what you mean ! Thank you for this! 

I believe the behavior you observe is due to the way runtime shiny is working: 

It will run a shiny app while rendering the doc to get the UI. So :

* It will first render the Rmd file using **rmarkdown** : This will run `get_runtime()` and the value of `rmarkdown.runtime` is set to shiny.
* Then it will run the shiny app based on the rendered UI element and the extracted server elements: this will be run like any other shiny apps. It is run using `shiny::shinyApp`. 

Also, as a comment, a shiny app built using Rmd files is a real shiny app. Your issue title seems to assume otherwise. 
The server code is not run from within a Rmd file or not once the app is launched. Once the shiny app is running, you won't get access to any R Markdown related variable or option you only get during a rendering process. This is because the running shiny apps is like any other. 

Can you share more on why you need to have such indicator available from inside the running shiny app ?

I think this would require to register this information as a classic R option (and not just a knitr's option) so that the shiny apps can access it. I am just curious why this information could be used for. 🤔

**Comment by cderv** [comment]
Is there anything else to look into or for us to do on this ? 
Re-reading the above with fresh eyes, it seems to me the behavior is the expected one: 
* The document will be rendered with knitr evaluating the chunk, something similar to a classic rmarkdown rendering where all chunks are evaluated sequentially into the same environment.
* They a shiny app will be run as any standard shiny app using `shiny::runApp` of the app created using `shiny::shinyApp`
https://github.com/rstudio/rmarkdown/blob/745c1477d02708938febb7065b85b552c7dc420f/R/shiny.R#L174-L178

So I believe this is a real Shiny, but the UI and Server are built from a R Markdown document content. For the shiny app in question to know that it has been started during a `rmarkdown::run()` it would require to set a state somewhere for the shiny app to retrieve it. If we go this road, knowing more about the use would help. 

I'll leave this open, but this requires more information so that something is planned for it. 

Thanks
